### ISSUE TITLE
[Access] Relay messages to unstaked network directly

### ISSUE THREAD

**Issue by smnzhu** [onflow/flow-go#1106]
This issue is to address the following concern raised about allowing unstaked nodes to connect to multiple upstream staked nodes:

> Staked nodes will be relaying duplicate information to the network, because on each new block proposal every staked nodes will relay it to the unstaked network. This means that an unstaked node which is connected to 2 upstream AN's will receive (and gossip-forward) every block proposal twice.

Currently, we are implementing our own relay engine to relay messages received on the staked network to the unstaked network: 

https://github.com/onflow/flow-go/blob/master/engine/access/relay/engine.go#L107-L108

However, this calls `Publish` on the libp2p `pubsub.Topic`:

https://github.com/onflow/flow-go/blob/master/network/p2p/libp2pNode.go#L487

Which creates a new libp2p [`Message`](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pb/rpc.pb.go#L143-L153) with the node's own ID / signature:

https://github.com/libp2p/go-libp2p-pubsub/blob/master/topic.go#L224-L234

In order to take advantage of libp2p's [built-in message deduplication](https://docs.libp2p.io/concepts/publish-subscribe/#:~:text=peers%20remember%20a%20list%20of%20recently%20seen%20messages.%20this%20lets%20peers%20act%20upon%20a%20message%20only%20the%20first%20time%20they%20see%20it%20and%20ignore%20retransmissions%20of%20already%20seen%20messages.), we need to relay the libp2p [`Message`](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pb/rpc.pb.go#L143-L153) unchanged.

Code pointers:

* [Forwarding a message received from another peer](https://docs.libp2p.io/concepts/publish-subscribe/#:~:text=similarly%2C%20when%20a%20peer%20receives%20a%20new%20message%20from%20another%20peer%2C%20it%20stores%20the%20message%20and%20forwards%20a%20copy%20to%20all%20other%20full-message%20peers%20it%20is%20connected%20to%3A): https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L1025-L1035
* The deduplication mentioned above: https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L1076-L1089
* Where the message is actually sent back out to peers: https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L1127
* GossipSub's implementation of `Publish`: https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L943

Instead of using the relay engine, what we probably need to do is to call `GossipSubRouter.Publish` directly to publish the message onto the unstaked network.

This will mean that we need some

**Comment by vishalchangrani** [comment]
I believe this is an improvement not only for the unstaked side of the network but also on the staked side of the network where multiple consensus nodes are publishing the same blocks.

**Comment by smnzhu** [comment]
So after some digging, it appears that it's actually not possible to access the `GossibSubRouter` from our code, since it is private. So there is no way to call its `Publish` method.

I'll keep looking as maybe there is a workaround, but for now it looks like maybe we have to find another way.

**Comment by smnzhu** [comment]
So I've done some more digging, and it turns out we won’t be able to access the `GossipSubRouter` since it is a private field. Thus, we can't directly relay messages by calling `GossipSubRouter` as hoped.

However, I realized that we may be approaching this completely wrong in the first place.

First of all, on the staked access node, we're creating two entirely separate instances of `Network`, which in turn creates two entirely distinct libp2p `Node`s

- [https://github.com/onflow/flow-go/blob/master/cmd/access/staked_access_node_builder.go#L77-L90](https://github.com/onflow/flow-go/blob/master/cmd/access/staked_access_node_builder.go#L77-L90)
- [https://github.com/onflow/flow-go/blob/master/network/p2p/middleware.go#L151](https://github.com/onflow/flow-go/blob/master/network/p2p/middleware.go#L151)

I'm guessing the reason it's implemented this way is because conceptually we are creating two "separate" networks, but in fact I'm not sure we have to explicitly separate them this way.

Here is a method that allows us to register a topic validator:

[https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L1297](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L1297)

The validators are run *before* any message is sent or gossiped-forward on the network

- [https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L1083-L1085](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L1083-L1085)
- [https://github.com/libp2p/go-libp2p-pubsub/blob/master/validation.go#L228-L244](https://github.com/libp2p/go-libp2p-pubsub/blob/master/validation.go#L228-L244)

What we could to instead, is the following:

- For all staked AN's, we register a validator that checks if the message originated from a node who is in the protocol state (ie a staked node). If it is, then we can allow the validator to pass.
    - What this implies is, that messages from the staked network will be gossiped-forward to the unstaked nodes
- If the message did not original from a node that we recognize from the protocol state, we fail the validation. *At the same time*, we intercept the message (since it must be from an unstaked node) and process it as we should.
    - This implies that messages from the public / unstaked network won't be gossiped-forward to the staked network.
    - At the same time, the staked AN's themselves will receive the messages from the unstaked network, and can process them however it needs (respond to sync requests, etc.)

The reason this should work is because unstaked nodes can only bootstrap from staked AN's and other unstaked nodes. In other words, the staked AN's are the only things connecting the staked network with the public network, and the two would be completely disconnected otherwise. This is ensured by connection gating on all of the staked nodes. Therefore, all traffic between the public and private networks has to pass through them.

One thing I'm not sure about is how this would work with DHT. Ideally, we would want to ensure that it is possible for the rest of the staked nodes to not participate in the DHT, so that unstaked nodes can't find out their addresses via the DHT. However, i believe it should be okay even if they do find out about the addresses because connection gating will prevent any messages from unstaked nodes to be gossiped by staked nodes.

**Comment by vishalchangrani** [comment]
The separation of the networks is intentional. We do not want any traffic on the staked network from nodes which are not staked.

**Comment by huitseeker** [comment]
Yes, but we have a Connection gater, don't we @vishalchangrani ?

**Comment by smnzhu** [comment]
Additional notes here: (Recording things here so I remember before putting everything into notion doc)

1. Wanted to verify that signature checking is done by libp2p **before** validators are called. In other words, by the time our validator is run, we know that the `From` field has been validated and we can take it at face value.
    * first, we check signing policy [here](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L1062-L1066). This ensures that if signing is required, that a signature must be present.
    * Then, [signature validation](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L1062-L1066) is done **before** any of [our validators](https://github.com/libp2p/go-libp2p-pubsub/blob/master/validation.go#L295-L342) are called.
2. [Here](https://github.com/libp2p/go-libp2p-pubsub/blob/master/topic.go#L224-L234) is where the `From` field is populated. [Here](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L153-L155) is the field definition.
3. **Important**: We must ensure that [sign policy](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L243) is never changed from the "strict" setting, for all of the nodes.
4. Here are the acceptable types we can use to implement a validator: https://github.com/libp2p/go-libp2p-pubsub/blob/master/validation.go#L162-L170

**Comment by smnzhu** [comment]
Another note:

Since we would be intercepting messages from unstaked nodes at the gossip layer, they would never make it to the engines on their own. In other words, they wouldn't go through the normal processing flow of messages received on the staked network (from libp2p -> our network layer -> engine). 

Instead, we will need some way of injecting that message back into the processing flow once it's been validated, whether by directly passing it to the network layer or directly passing it to an engine.
